pkg/kubelet/cm/topologymanager/policy.go

```diff
@@ -251,41 +270,44 @@ func compareHints(bestNonPreferredAffinityCount int, current *TopologyHint, cand
 	if candidate.NUMANodeAffinity.Count() < current.NUMANodeAffinity.Count() {
 		return current
 	}
+
 	// Case 3cc
-	if candidate.NUMANodeAffinity.IsNarrowerThan(current.NUMANodeAffinity) {
-		return candidate
-	}
-	return current
+	return m.GetBetterHint(candidate, current)
+
 }
 
-func mergeFilteredHints(numaNodes []int, filteredHints [][]TopologyHint) TopologyHint {
-	// Set bestNonPreferredAffinityCount to help decide which affinity mask is
-	// preferred amongst all non-preferred hints. We calculate this value as
-	// the maximum of the minimum affinity counts supplied for any given hint
-	// provider. In other words, prefer a hint that has an affinity mask that
-	// includes all of the NUMA nodes from the provider that requires the most
-	// NUMA nodes to satisfy its allocation.
-	bestNonPreferredAffinityCount := maxOfMinAffinityCounts(filteredHints)
+func (m HintMerger) Merge() TopologyHint {
+	defaultAffinity := m.NUMAInfo.DefaultAffinityMask()
 
 	var bestHint *TopologyHint
-	iterateAllProviderTopologyHints(filteredHints, func(permutation []TopologyHint) {
+	iterateAllProviderTopologyHints(m.FilteredHints, func(permutation []TopologyHint) {
 		// Get the NUMANodeAffinity from each hint in the permutation and see if any
 		// of them encode unpreferred allocations.
-		mergedHint := mergePermutation(numaNodes, permutation)
+		mergedHint := mergePermutation(defaultAffinity, permutation)
 
 		// Compare the current bestHint with the candidate mergedHint and
 		// update bestHint if appropriate.
-		bestHint = compareHints(bestNonPreferredAffinityCount, bestHint, &mergedHint)
+		bestHint = m.compare(bestHint, &mergedHint)
 	})
 
 	if bestHint == nil {
-		defaultAffinity, _ := bitmask.NewBitMask(numaNodes...)
 		bestHint = &TopologyHint{defaultAffinity, false}
 	}
 
 	return *bestHint
 }
 
+func (m HintMerger) GetBetterHint(h1, h2 *TopologyHint) *TopologyHint {
```

Comment:
To make it work as before we need to return `current` hint if both `bitmasks` are the same, so I guess we can do something like:
```go
		CompareNUMAAffinityMasks: func(candidate, current *TopologyHint) *TopologyHint {
			// If current and candidate bitmask is the same, prefer current hint.
			if candidate.NUMANodeAffinity.IsEqual(current.NUMANodeAffinity) {
				return current
			}

			best := compareMasks(candidate.NUMANodeAffinity, current.NUMANodeAffinity)
			if best.IsEqual(candidate.NUMANodeAffinity) {
				return candidate
			}
			return current
		},
```

Reply:
Sure. Makes sense. Maybe even define the whole function outside of the struct definition as the following to clean it up a bit:
```
func NewHintMerger(numaInfo *NUMAInfo, filteredHints [][]TopologyHint, policyName string, opts PolicyOptions) HintMerger {
	compareNumaAffinityMasks := func(current, candidate *TopologyHint) *TopologyHint) {
		// If current and candidate bitmasks are the same, prefer current hint.
		if candidate.NUMANodeAffinity.IsEqual(current.NUMANodeAffinity) {
			return current
		}

		// Otherwise compare the hints, based on the policy options provided
		var best *TopologyHint
		if (policyName != PolicySingleNumaNode) && opts.PreferClosestNUMA {
			best = numaInfo.Closest(current.NUMANodeAffinity, candidate.NUMANodeAffinity)
		} else {
			best = numaInfo.Narrowest(current.NUMANodeAffinity, candidate.NUMANodeAffinity)
		}
		if best.IsEqual(current.NUMANodeAffinity) {
			return current
		}
		return candidate
	}

	return HintMerger{
		NUMAInfo:                      numaInfo,
		Hints:                         hints,
		BestNonPreferredAffinityCount: maxOfMinAffinityCounts(hints),
		CompareNUMAAffinityMasks:      compareNumaAffinityMasks,
	}
}
```